Add IPsec tunnel state metric for IKE Child SA establishment

pperiyasamy opened the pull request:
This PR registers metric adds `ovnkube_controller_ipsec_tunnel_ike_child_sa_state` metric which reflects IKE Child SA establishment state for all the configured IPsec tunnels via OVN. When all tunnels has IKE Child SAs established, then metric is set to 1, otherwise it will be reset to 0. The implementation is done by listening to xfrm netlink change events for xfrm state and policy.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- New Features
  - Optional IPsec monitoring via feature flag (--enable-ipsec) with a metric showing tunnel health.
  - Node-side monitoring now observes tunnel and IPSec state to update metrics.

- Helm
  - Added startup safeguard to remove stale pluto.pid for reliable restarts.
  - Renamed env var to OVN_ENABLE_IPSEC.
  - Removed v3_req extension from CSR generation.

- Tests & CI
  - New unit and e2e tests for IPsec metrics and recovery; CI matrix and prom2json install added.

- Chores
  - Kind script switched CSR handling from oc to kubectl.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

pperiyasamy committed:
Provide IPsec enable option for IC cluster

The IPsec enable option is currently not set for IC cluster,
This commit updates the scripts to configure ovn nbdb with
ipsec option when it's set while creating kind cluster.

Signed-off-by: Periyasamy Palanisamy <pepalani@redhat.com>

pperiyasamy committed:
Remove extensions parameter for CSR create command

When CSR is created with extensions parameter, the following error is thrown:

+ openssl req -new -text -extensions v3_req -addext subjectAltName=DNS:f442ae14-1b76-46e4-955e-e4e864f8fef3 -subj /C=US/O=ovnkubernetes/OU=kind/CN=f442ae14-1b76-46e4-955e-e4e864f8fef3 -key /etc/openvswitch/keys/ipsec-privkey.pem -out /etc/openvswitch/keys/ipsec-req.pem
Error adding request extensions defined via -addext
80A23BD0237F0000:error:0580008C:x509 certificate routines:X509at_add1_attr_by_NID:duplicate attribute:crypto/x509/x509_att.c:194:

This happens though openssl.cnf file doesn't include subjectAltName for v3_req section, so skipping
`-extensions v3_req` parameter while creating CSR. As expected by libreswan the subjectAltName is
also present in the certificate.

Signed-off-by: Periyasamy Palanisamy <pepalani@redhat.com>

pperiyasamy committed:
Add IPsec tunnel IKE Child SA state metric

This commit adds ovnkube_controller_ipsec_tunnel_ike_child_sa_state metric
which reflects IKE Child SA establishment state for all the configured IPsec
tunnels via OVN. When all tunnels has IKE Child SAs established, then metric
is set to 1, otherwise it will be reset to 0. The implementation is done by
listening to xfrm netlink change events for xfrm state and policy.

Signed-off-by: Periyasamy Palanisamy <pepalani@redhat.com>

martinkennelly commented on dist/images/ovnkube.sh:
```diff
@@ -981,6 +984,11 @@ local-nb-ovsdb() {
 
+ [[ "true" == "${ENABLE_IPSEC}" ]] && {
+    ovn-nbctl set nb_global . ipsec=true
+    echo "=============== nb-ovsdb ========== reconfigured for ipsec"
```
Why reconfigured ?

pperiyasamy replied:
this was needed for IC cluster, otherwise IPsec is not configured in nbdb.

martinkennelly commented on dist/templates/ovn-ipsec.yaml.j2:
```diff
@@ -101,7 +102,6 @@ spec:
             # Generate an SSL private key and use the key to create a certitificate signing request
             umask 077 && openssl genrsa -out /etc/openvswitch/keys/ipsec-privkey.pem 2048
             openssl req -new -text \
-                        -extensions v3_req \
```
Can you explain this ?

pperiyasamy replied:
as explained in the commit message, CSR creation is failed with `-extensions v3_req` parameter on fedora, so had to remove this, but this didn't harm in configuring IPsec with libreswan.

martinkennelly replied:
I am not a SME on this so i cannot say if this is ok or not.. if its not. Id add it to a new PR if its unrelated.

martinkennelly replied:
can we get some security specialist to review this? 
I cannot validate removing this arg. :/ 
I dont have experience so I cannot confirm if what you said is true.

martinkennelly replied:
this should be seperate PR i think - seems unrelated to mertrics.

martinkennelly commented on dist/templates/ovnkube-single-node-zone.yaml.j2:
```diff
@@ -299,6 +301,10 @@ spec:
           subPath: private
           readOnly: true
+        {% if enable_ipsec=="true" -%}
+        - mountPath: /var/run/pluto
```
Does that path exist upstream? What's this path for ?

pperiyasamy replied:
This is to get access to `/var/run/pluto/pluto.ctl` UDS file from `ovnkube-controller` container to run ipsec command to retrieve states from pluto process.

martinkennelly commented on go-controller/pkg/metrics/ovnkube_controller.go:
```diff
@@ -257,6 +261,16 @@ var metricIPsecEnabled = prometheus.NewGauge(prometheus.GaugeOpts{
+	Subsystem: MetricOvnkubeSubsystemController,
+	Name:      "ipsec_tunnel_state",
+	Help:      "Specifies whether IPSec Child SA is established for the Geneve tunnel(1) or not(0)"},
+	[]string{
```
You want to add an alarm too for this ? Well probably use it downstream so may as well add it upstream?

pperiyasamy replied:
I don't see much of IPsec deployments with u/s, so let's start with adding alert rule in d/s.

martinkennelly commented on go-controller/pkg/metrics/ovnkube_controller.go:
```diff
@@ -601,6 +615,14 @@ func UpdateEgressFirewallRuleCount(count float64) {
 }
 
+func RecordIPsecTunnelStateUpEvent(localIP, remoteIP string) {
+	metricIPsecTunnelState.WithLabelValues(localIP, remoteIP).Set(1)
```
If the values are always the same together.. Why have two time series ?

martinkennelly commented on go-controller/pkg/metrics/ovnkube_controller.go:
```diff
@@ -649,6 +671,49 @@ func ipsecMetricHandler(table string, model model.Model) {
+				}
+				if !nbGlobal.Ipsec {
+					// IPsec is not enabled, so keep checking for ipsec after wait time.
+					time.Sleep(ipSecTunnelStatusQueryWaitTime)
```
This a go routine. Youre blocking it here and it cannot exit. Also, theres no way for this go routine to say I'm done, you can exit, we usually use wait groups. Is it possible ?

Why not expose its true value directly instead of waiting around for it to be true ?

pperiyasamy replied:
oh! yes, now gracefully exiting using wait group.

we don't need to report this metric if ipsec is not enabled by user, hence this check is needed.

martinkennelly commented on go-controller/pkg/util/ipsec.go:
```diff
@@ -0,0 +1,31 @@
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+func GetAllEstablishedIPsecTunnels() (sets.Set[string], error) {
```
Is there an ipsec client to get this info ?

pperiyasamy replied:
AFAIK we have only `ipsec` cli to connect with pluto daemon, do you see any problem with using it ?

martinkennelly commented on go-controller/pkg/util/ovs.go:
```diff
@@ -249,6 +251,10 @@ func SetExecWithoutOVS(exec kexec.Interface) error {
 		if err != nil {
 			return err
 		}
+		runner.ipsecPath, err = exec.LookPath(ipsecCommand)
```
Should we only do this when ipsec is enabled ?

pperiyasamy replied:
I think it's ok to initialize ipsec command by default (ovnk image have ipsec binary by default).

martinkennelly commented on go-controller/pkg/util/ovs.go:
```diff
@@ -895,3 +907,50 @@ func GetOVSPortPodInfo(hostIfName string) (bool, string, string, error) {
+	// {"data":[["ovn-203886-0",["map",[["csum","true"],["key","flow"],["local_ip","10.89.0.3"],["remote_ip","10.89.0.4"],["remote_name","203886dd-d6e1-4b48-bdd9-0a6bbfc6acdc"]]]],
+	//          ["ovn-4e658c-0",["map",[["csum","true"],["key","flow"],["local_ip","10.89.0.3"],["remote_ip","10.89.0.2"],["remote_name","4e658c65-a43f-4bb9-a36f-723ef1d96a66"]]]]],
+	//  "headings":["name","options"]}
+	data := result["data"].([]interface{})
```
Should you check there's data before casting ?

pperiyasamy replied:
yes @martinkennelly , this function needed rework on parsing command output and proper error handling, done.

martinkennelly replied:
waiting on this

martinkennelly replied:
unresolved this peri

martinkennelly commented on go-controller/pkg/util/ovs.go:
```diff
@@ -895,3 +907,50 @@ func GetOVSPortPodInfo(hostIfName string) (bool, string, string, error) {
+	data := result["data"].([]interface{})
+	for _, item := range data {
+		entry := item.([]interface{})
+		name := entry[0].(string)
```
This code scares me with all the casting

pperiyasamy replied:
look at the new function, hope that is fine with you.

martinkennelly commented on go-controller/pkg/util/ovs.go:
```diff
@@ -895,3 +907,50 @@ func GetOVSPortPodInfo(hostIfName string) (bool, string, string, error) {
+	for _, item := range data {
+		entry := item.([]interface{})
+		name := entry[0].(string)
+		options := entry[1].([]interface{})
```
How do we know it has two entries ?

pperiyasamy replied:
took care of it.

martinkennelly reviewed (commented):
Needs more testing of accessing the db and parsing the tunnels. Also, testing by bringing the state up and down to exercise that code. You may have to fake some things 

Also, parsing the db output when looking for the tunnels is hard to read. Is there a better way ? Unmarall into interface list doesnt seem like best way to go. Wdyt..

Is there any ipsec client instead of using ipsec binary and parsing result ?

martinkennelly commented on go-controller/pkg/metrics/ovnkube_controller.go:
```diff
@@ -601,6 +615,14 @@ func UpdateEgressFirewallRuleCount(count float64) {
 }
 
+func RecordIPsecTunnelStateUpEvent(localIP, remoteIP string) {
+	metricIPsecTunnelState.WithLabelValues(localIP, remoteIP).Set(1)
```
This won't be accepted downstream because its cardinality isnt bounded.

pperiyasamy replied:
yes @martinkennelly , understand your concern. the number of metrics will go out of proportion in a scaled cluster, should we generate this metric only if it's asked by the user ? or shall we report only problematic ipsec connections ?

pperiyasamy reviewed (commented):
> Needs more testing of accessing the db and parsing the tunnels. Also, testing by bringing the state up and down to exercise that code. You may have to fake some things

added required unit tests for the metric. it needed some mocking around some objects, PTAL.

> Also, parsing the db output when looking for the tunnels is hard to read. Is there a better way ? Unmarall into interface list doesnt seem like best way to go. Wdyt..

yes @martinkennelly , refactored the parsing logic in `listGeneveTunnels` function now. 

> Is there any ipsec client instead of using ipsec binary and parsing result ?

I'm not aware of any reliable ipsec client other than ipsec binary, are you having someing in your mind ?

martinkennelly commented on dist/templates/ovn-ipsec.yaml.j2:
```diff
@@ -258,8 +258,8 @@ spec:
           name: host-cni-netd
-        - mountPath: /var/run/openvswitch
-          name: host-var-run-ovs
+        - mountPath: /var/run
```
what else are you consuming from run? why not list it specifically instead of the whole dir?

martinkennelly commented on go-controller/pkg/config/config.go:
```diff
@@ -1327,6 +1332,12 @@ var MetricsFlags = []cli.Flag{
 		Usage:       "Enables metrics related to scaling",
 		Destination: &cliConfig.Metrics.EnableScaleMetrics,
 	},
+	&cli.DurationFlag{
```
we dont normally expose this - whats the use case?

pperiyasamy replied:
we may configure this value lesser than 10s, it may be helpful in detecting ipsec tunnel flaps if that happens within 10 secs, could be useful for troubleshooting d/s CI disruptive events if that is caused by ipsec tunnel establishment failures.

martinkennelly commented on go-controller/pkg/metrics/ipsec.go:
```diff
@@ -0,0 +1,33 @@
+	lines := strings.Split(string(stdout), "\n")
+	for _, line := range lines {
+		if strings.Contains(line, "ESTABLISHED_CHILD_SA") {
+			matches := regexp.MustCompile(`"([^"]*)"`).FindAllStringSubmatch(line, -1)
```
maybe write a comment on the expression?

pperiyasamy replied:
done.

martinkennelly replied:
hmm - matching on things just having quotes doesnt sound great. I know its unlikely the output changes but it seems super weak to match on all quoted words. I thought we'd only expect one at least but i see you dont limit the number of matches. I am hoping theres json output for this cli.

pperiyasamy replied:
there is no json output option for ipsec commands for now.

martinkennelly commented on go-controller/pkg/metrics/ipsec.go:
```diff
@@ -0,0 +1,33 @@
+		if strings.Contains(line, "ESTABLISHED_CHILD_SA") {
+			matches := regexp.MustCompile(`"([^"]*)"`).FindAllStringSubmatch(line, -1)
+			for _, m := range matches {
+				ipsecTunnels.Insert(strings.Fields(m[1])...)
```
shouldnt we double check that the len is greater than 1 so we dont panic and bring down the control plane?

martinkennelly commented on go-controller/pkg/metrics/ovnkube_controller.go:
```diff
@@ -254,6 +255,14 @@ var metricIPsecEnabled = prometheus.NewGauge(prometheus.GaugeOpts{
+	Subsystem: types.MetricOvnkubeSubsystemController,
+	Name:      "ipsec_tunnel_state",
+	Help:      "Specifies whether IPSec Child SA is established for the Geneve tunnel(1) or not(0)"},
+	[]string{"remote_ip"},
```
unbounded cardinality :( we cannot downstream this commit.

pperiyasamy replied:
the intention is have a metric for every geneve tunnel, so what is your thinking ? just to have a single metric which conveys overall tunnel status ?

martinkennelly replied:
ya, unless you can safely say the cardinality is < 14.

martinkennelly commented on go-controller/pkg/metrics/ovs.go:
```diff
@@ -920,3 +922,76 @@ func registerOvsMetrics(ovsDBClient libovsdbclient.Client, metricsScrapeInterval
+		return geneveTunnels, fmt.Errorf("failed to parse json output for Geneve tunnels, err: %v", err)
+	}
+	for _, entry := range ovsResult.Data {
+		klog.Infof("data: %v", entry)
```
remove

pperiyasamy replied:
yes, this was just filling up ovnkube-controller logs in the d/s testing. removed it.

martinkennelly commented on go-controller/pkg/metrics/ovs.go:
```diff
@@ -920,3 +922,76 @@ func registerOvsMetrics(ovsDBClient libovsdbclient.Client, metricsScrapeInterval
+	for _, entry := range ovsResult.Data {
+		klog.Infof("data: %v", entry)
+		if len(entry) != 4 {
+			continue
```
would this ever happen? can we log an error. Trying to think of scenario where api changes and we want some visibility.

pperiyasamy replied:
hope the output is always fixed, anyway adding error msg now.

martinkennelly commented on go-controller/pkg/metrics/ovs.go:
```diff
@@ -920,3 +922,76 @@ func registerOvsMetrics(ovsDBClient libovsdbclient.Client, metricsScrapeInterval
+		return geneveTunnels, fmt.Errorf("failed to parse json output for Geneve tunnels, err: %v", err)
+	}
+	for _, entry := range ovsResult.Data {
+		klog.Infof("data: %v", entry)
```
can you give a doc text of a sample output so i can review below?

pperiyasamy replied:
sure, done.

martinkennelly commented on go-controller/pkg/metrics/ovnkube_controller.go:
```diff
@@ -619,6 +653,56 @@ func ipsecMetricHandler(table string, model model.Model) {
+				nbGlobal := &nbdb.NBGlobal{}
+				nbGlobal, err := libovsdbops.GetNBGlobal(nbClient, nbGlobal)
+				if err != nil {
+					klog.Errorf("Error in getting the NBGlobal row from Northbound db, err: %v", err)
```
do you really want to exit here? i thought youd only want to exit on shut down (stopCh closed)?  `MontiorIPsecTunnelsState` is exec only once.

martinkennelly commented on go-controller/pkg/metrics/ovnkube_controller.go:
```diff
@@ -619,6 +653,56 @@ func ipsecMetricHandler(table string, model model.Model) {
+// MontiorIPsecTunnelsState will register a labeled metric for the ipsec tunnel status. It will keep
+// checking ipsec established status for each Geneve tunnel towards other remote node for every 5s
+// and update the metric with tunnel up or down status.
+func MontiorIPsecTunnelsState(stopChan <-chan struct{}, wg *sync.WaitGroup, nbClient libovsdbclient.Client,
```
did you look into adding event handles for this instead of polling? just curious.

pperiyasamy replied:
yes @martinkennelly , seems like `netlink.SubscribeAt(nl.NetnsNone(), nl.NetnsNone(), unix.NETLINK_XFRM, 0xffffffff)` can be used here, will try this out.

huiran0826 committed:
add ipsec e2e metric case

e2e case to test IPSec metrics

Signed-off-by: Huiran Wang <huirwang@redhat.com>

pperiyasamy commented on go-controller/pkg/metrics/ovnkube_controller.go:
```diff
@@ -619,6 +647,112 @@ func ipsecMetricHandler(table string, model model.Model) {
 
+// MontiorIPsecTunnelsState will register a labeled metric for the ipsec tunnel status. It will subscribe
+// to xfrm netlink events and update the metric with the tunnel up or down status.
+func MontiorIPsecTunnelsState(stopChan <-chan struct{}, wg *sync.WaitGroup, nbClient libovsdbclient.Client,
```
yes, fixed it.

pperiyasamy commented:
/retest-failed

pperiyasamy committed:
Remove stale pluto.pid if pluto process is not running

It fixes ovn-ipsec pod crash when it is restarted.

Signed-off-by: Periyasamy Palanisamy <pepalani@redhat.com>

martinkennelly commented on dist/images/ovnkube.sh:
```diff
@@ -996,6 +999,11 @@ local-nb-ovsdb() {
   ovn-nbctl set NB_Global . options:name=${K8S_NODE}
 
+ [[ "true" == "${ENABLE_IPSEC}" ]] && {
+    ovn-nbctl set nb_global . ipsec=true
```
note to reviewers; set -e is enable so this would fail and exit if theres an issue setting this.

martinkennelly reviewed (commented):
@pperiyasamy is there any upstream jobs showing ipsec enabled disabled? I dont remember it.

martinkennelly commented on go-controller/pkg/util/ipsec.go:
```diff
@@ -0,0 +1,31 @@
+	ipsecTunnels := sets.Set[string]{}
+	stdout, stderr, err := RunIPsec("showstates")
+	if err != nil {
+		return ipsecTunnels, fmt.Errorf("failed to retrieve ipsec states %v: %v", stderr, err)
```
nit;; stderr can come right after string "states " might be confusing

pperiyasamy replied:
changes the error message accordingly.

martinkennelly commented on go-controller/pkg/util/ipsec.go:
```diff
@@ -0,0 +1,31 @@
+		if strings.Contains(line, "ESTABLISHED_CHILD_SA") {
+			matches := regexp.MustCompile(`"([^"]*)"`).FindAllStringSubmatch(line, -1)
+			for _, m := range matches {
+				ipsecTunnels.Insert(strings.Fields(m[1])...)
```
should you check that m is len 2 or greater before indexing otherwise panic

martinkennelly commented on dist/templates/ovn-ipsec.yaml.j2:
```diff
@@ -230,6 +230,12 @@ spec:
           # Workaround for https://github.com/libreswan/libreswan/issues/373
           ulimit -n 1024
 
+          # Remove stale pluto.pid if no pluto process is running
```
who writes this pid file? does it mean pluto exited early and didnt remove its pid file or ?

pperiyasamy replied:
ye, this creates pluto.pid file https://github.com/ovn-kubernetes/ovn-kubernetes/blob/master/dist/templates/ovn-ipsec.yaml.j2#L239 and it gets killed upon sigterm signal https://github.com/huiran0826/ovn-kubernetes/blob/ipsec-metric-e2e/dist/templates/ovn-ipsec.yaml.j2#L203.

martinkennelly commented on dist/templates/ovn-ipsec.yaml.j2:
```diff
@@ -230,6 +230,12 @@ spec:
+          # Remove stale pluto.pid if no pluto process is running
+          if [ -f /var/run/pluto/pluto.pid ] && ! pgrep -F /var/run/pluto/pluto.pid > /dev/null 2>&1; then
+            echo "Removing stale pluto.pid"
+            rm -f /var/run/pluto/pluto.pid
```
sorry not familiar with this code but i see youre removing the pid file in two places - here and the daemonset - can you explain more why you did that?

pperiyasamy replied:
when `ovn-ipsec` pod is restarted, then pluto is killed, but its pid file on the host is not deleted timely before new ovn-ipsec container comes up, so it's failing to spin up new pluto process because old pid file is still present. so this code deletes stale pid file if present.

The daemonset.yaml is a template file for helm based deployment, so we needed this change at two places.

martinkennelly replied:
I see - thanks. Are you only seeing the stale pid if the node is force rebooted / power off scenarios or?

huiran0826 committed:
add ipsec lane

add ci lane for ipsec

Signed-off-by: Huiran Wang <huirwang@redhat.com>

pperiyasamy commented on go-controller/pkg/util/ovs.go:
```diff
@@ -908,3 +920,50 @@ func GetOVSPortPodInfo(hostIfName string) (bool, string, string, error) {
+	var result map[string]interface{}
+	err = json.Unmarshal([]byte(stdout), &result)
+	if err != nil {
+		return geneveTunnels, fmt.Errorf("failed to parse json output for Geneve tunnels, err: %v", err)
```
done.

pperiyasamy commented on go-controller/pkg/metrics/ovnkube_controller.go:
```diff
@@ -619,6 +641,49 @@ func ipsecMetricHandler(table string, model model.Model) {
+// MontiorIPsecTunnelsState will register a labeled metric for the ipsec tunnel status. It will keep
+// checking ipsec established status for each Geneve tunnel towards other remote node for every 5s
+// and update the metric with tunnel up or down status.
+func MontiorIPsecTunnelsState(stopChan <-chan struct{}, nbClient libovsdbclient.Client) {
```
agree, but this is not the case anymore with new xfrm event handler implementation. it's being returned gracefully after proper cleanup.

pperiyasamy commented on go-controller/pkg/metrics/ovnkube_controller.go:
```diff
@@ -254,6 +258,16 @@ var metricIPsecEnabled = prometheus.NewGauge(prometheus.GaugeOpts{
+	Name:      "ipsec_tunnel_state",
+	Help:      "Specifies whether IPSec Child SA is established for the Geneve tunnel(1) or not(0)"},
+	[]string{
+		"local_ip", "remote_ip",
```
this is not unbounded cardinality anymore with commit https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5201/commits/e499f32c05696cbe726657b7318d663fd40b54a7.

martinkennelly commented on dist/images/daemonset.sh:
```diff
@@ -106,6 +106,7 @@ IN_UPGRADE=
 # northd-backoff-interval, in ms
 OVN_NORTHD_BACKOFF_INTERVAL=
 OVN_OBSERV_ENABLE="false"
+ENABLE_IPSEC="false"
```
this seems different to adding metrics and probably should be seperate PR.

huiran0826 replied:
@pperiyasamy  I added my commits to your PR.

martinkennelly commented on dist/templates/ovnkube-zone-controller.yaml.j2:
```diff
@@ -442,6 +446,12 @@ spec:
+      - name: run-pluto
+        hostPath:
+          path: /var/run/pluto
+          type: DirectoryOrCreate
```
would it be already created prio to ovnk coming up? can we fail the directory doesnt exist instead of creating?

martinkennelly commented on go-controller/pkg/metrics/ipsec.go:
```diff
@@ -0,0 +1,50 @@
+			matches := regexp.MustCompile(`"([^"]*)"`).FindAllStringSubmatch(line, -1)
+			for _, match := range matches {
+				if len(match) != 2 {
+					klog.Warningf("Invalid ipsec tunnel entry %s for ipsec showstates command, skip processing it", line)
```
id say put the offended line at the end and not the middle of the message as it may confuse readers.

martinkennelly commented on go-controller/pkg/metrics/ipsec.go:
```diff
@@ -0,0 +1,44 @@
+				// match always contains two elements.
+				// match[0]: The full text that matched the regex, for example "\"ovn-60abc6-0-in-1\"".
+				// match[1]: The captured group inside the match, i.e. the part inside the parentheses, for example "ovn-60abc6-0-in-1".
+				ipsecTunnels.Insert(strings.Fields(match[1])...)
```
the bot seems right though no? why do we need strings.Fields if we dont expect spaces.

martinkennelly commented on go-controller/pkg/metrics/ipsec.go:
```diff
@@ -0,0 +1,50 @@
+			}
+		}
+	}
+	return ipsecTunnels, nil
```
at this point, should we have some validation that theres at least one tunnel?

martinkennelly commented on go-controller/pkg/metrics/ipsec_test.go:
```diff
@@ -0,0 +1,223 @@
+		ginkgo.By("clean up resources for the test")
+		close(stopChan)
+		wg.Wait()
+		err := netlink.XfrmStateDel(getBaseState())
```
need root? should this test be producted by OnSupportedEnv

martinkennelly replied:
yes, peri .. wow what a good bot!!

martinkennelly replied:
@coderabbitai do you have persistent storage or config option so i can tell you to perform this check for all new unit tests that come into this repo?

martinkennelly commented on go-controller/pkg/metrics/ipsec_test.go:
```diff
@@ -0,0 +1,223 @@
+			ipsec := NewFakeIPsecClient(ipsecCmdOutput)
+			MonitorIPsecTunnelsState(stopChan, wg, ovsVsctl.FakeCall, ipsec.FakeCall)
+			time.Sleep(2 * time.Second) // Allow some time for the goroutine to start
+			err := netlink.XfrmStateAdd(getBaseState())
```
can you explain what this is and why do it?

martinkennelly commented on go-controller/pkg/metrics/ovnkube_controller.go:
```diff
@@ -619,6 +647,106 @@ func ipsecMetricHandler(table string, model model.Model) {
 
+// MonitorIPsecTunnelsState will register a labeled metric for the ipsec tunnel status. It will subscribe
+// to xfrm netlink events and update the metric with the tunnel up or down status.
+func MonitorIPsecTunnelsState(stopChan <-chan struct{}, wg *sync.WaitGroup, ovsVsctl ovsClient, ipsec ipsecClient) {
```
would it be easier to unit test if you launched the go routine in a seperate func? then you wouldnt need your sleep in your unit test to wait for goroutine to come up.

martinkennelly replied:
no. what i meant is its hard to unit test funcs when they launch go routines - theres no option to not launch a go routine if you wanted to test the func serially - in-line - which doesnt sound good for a unit test. Increases chances of go routine leak too between tests.

martinkennelly commented on go-controller/pkg/metrics/ovnkube_controller.go:
```diff
@@ -619,6 +647,106 @@ func ipsecMetricHandler(table string, model model.Model) {
+			nl.XFRM_MSG_NEWPOLICY, nl.XFRM_MSG_DELPOLICY, nl.XFRM_MSG_UPDPOLICY})
+		if err != nil {
+			klog.Errorf("Error creating xfrm Mcast groups, err: %v", err)
+			return
```
give up forever if we fail here? no retry?

martinkennelly commented on go-controller/pkg/metrics/ovnkube_controller.go:
```diff
@@ -619,6 +647,106 @@ func ipsecMetricHandler(table string, model model.Model) {
+			return
+		}
+		if stopChan != nil {
+			wg.Add(1)
```
this should always be added syncroneous not async within go routine - race here

martinkennelly commented on go-controller/pkg/metrics/ovnkube_controller.go:
```diff
@@ -619,6 +647,106 @@ func ipsecMetricHandler(table string, model model.Model) {
+			}()
+		}
+		for {
+			msgs, from, err := sock.Receive()
```
how does this end if stop signal received? Its a bit strange and hard to read that youve another go routine above and then this for loop below that seems to do its own thing.. i see itll probably emit an info log but this is weird.. i think you should use the common pattern we do - switch statement

martinkennelly reviewed (commented):
a lot to review here tbh - i expected just metrics stuff but theres also other things in here it seems - making the PR harder to reason and review imo. Ill continue tomorrow.

martinkennelly commented on dist/templates/ovn-ipsec.yaml.j2:
```diff
@@ -66,6 +66,7 @@ spec:
       serviceAccountName: ovnkube-node
       hostNetwork: true
       dnsPolicy: Default
+      {{ "hostPID: true" if ovn_unprivileged_mode=="no" }}
```
i see - it took a long time for me to figure out the meaning of this unpriv mode. change seems fine.

pperiyasamy commented:
The xfrm event subscription is failing in non-ic deployments due to permission issue and causing kind cluster bring up to fail. investigating...

pperiyasamy committed:
Conditionally collect IKE child SA establishment metrics

Only collect IPsec IKE child SA establishment metrics when IPsec is
enabled by passing the enable-ipsec flag to ovnkube-controller or
ovnkube-node container.

This avoids unnecessary metric collection overhead when IPsec is
not in use.

Signed-off-by: Periyasamy Palanisamy <pepalani@redhat.com>

pperiyasamy commented:
> The xfrm event subscription is failing in non-ic deployments due to permission issue and causing kind cluster bring up to fail. investigating...

There are permission issues in subscribing xfrm events in ovnkube-node pod in non-ic and multizone IC deployments. it's also needed to collect this metric as a node level metric, fixed that as well.

pperiyasamy commented on go-controller/pkg/metrics/mocks/gauge.go:
```diff
@@ -9,56 +12,143 @@ import (
 	panic("unimplemented")
 }
 
-func (GaugeMock) Write(*io_prometheus_client.Metric) error {
+func (h *GaugeMock) Write(*io_prometheus_client.Metric) error {
 	panic("unimplemented")
 }
 
-func (GaugeMock) Describe(chan<- *prometheus.Desc) {
-	panic("unimplemented")
+func (h *GaugeMock) Describe(chan<- *prometheus.Desc) {
 }
 
-func (GaugeMock) Collect(chan<- prometheus.Metric) {
+func (h *GaugeMock) Collect(chan<- prometheus.Metric) {
 	panic("unimplemented")
 }
```
I decide to keep as it is, because no unit tests are impacted by this. we can revisit it later when needed.

pperiyasamy commented on go-controller/pkg/metrics/ipsec_test.go:
```diff
@@ -0,0 +1,240 @@
+		gomega.Expect(err).NotTo(gomega.HaveOccurred(), "monitor ipsec tunnel state should not fail")
+	})
+
+	ginkgo.AfterEach(func() {
+		ginkgo.By("clean up resources for the test")
+		close(stopChan)
+		wg.Wait()
+		// Clean up ip xfrm state entry after every test.
+		err := netlink.XfrmStateDel(getBaseState())
+		gomega.Expect(err).NotTo(gomega.HaveOccurred(), "should not return an error for ip xfrm state entry cleanup")
+	})
```
we don't need to Unregister the metric for every test, take a look new PR update.

martinkennelly reviewed (commented):
wait for refactor - finished at commit 3201025343d71a3ff62a3e4b5da48cbe52beca04

martinkennelly commented:
Needs simplifaction peri - its too complex as is imo and its hard to read the code imo.
